src/org/jgroups/raft/blocks/LockService.java

```diff
@@ -0,0 +1,808 @@
+				if (handler != null) try {
+					handler.accept(this);
+				} catch (Throwable e) {
+					log.error("Error occurred on unlock handler", e);
+				}
+			} else if (curr != NONE && acquirers.get() == 0) {
+				status = curr;
+				var handler = lockHandler;
+				if (handler != null) try {
+					handler.accept(this);
+				} catch (Throwable e) {
+					log.error("Error occurred on lock handler", e);
+				}
+			} else if (prev == WAITING) {
+				delegate.lock();
+				try {
+					if (status == WAITING) {
+						status = curr;
+						notWaiting.signalAll();
+					}
+				} finally {
+					delegate.unlock();
+				}
+			}
+		}
+	}
+
+	private static <T> T rethrow(Throwable e) {
+		if (e instanceof RaftException) throw (RaftException) e;
+		if (e instanceof CompletionException) {
+			Throwable cause = e.getCause();
+			throw cause != null ? new RaftException(e) : (CompletionException) e;
+		}
+		if (e instanceof ExecutionException) throw new RaftException(e.getCause());
+		if (e instanceof TimeoutException) throw new RaftException("Execute command timeout", e);
+		throw new RaftException("Unknown exception", e);
+	}
+
+	private static void writeInt(int value, DataOutput out) {
+		try {
+			for (; (value & ~0x7F) != 0; value >>>= 7) {
+				out.writeByte(0x80 | (value & 0x7F));
+			}
+			out.writeByte(value);
+		} catch (IOException e) {
+			throw new RaftException("Fail to write", e);
+		}
+	}
+
+	private static int readInt(DataInput in) {
+		try {
+			int v = in.readByte(); if (v >= 0) return v;
+			if ((v ^= in.readByte() << 7)  <  0) return v ^ 0xFFFFFF80;
+			if ((v ^= in.readByte() << 14) >= 0) return v ^ 0x00003F80;
+			if ((v ^= in.readByte() << 21) <  0) return v ^ 0xFFE03F80;
+			return v ^ in.readByte() << 28 ^ 0x0FE03F80;
+		} catch (IOException e) {
+			throw new RaftException("Fail to read", e);
+		}
+	}
+
+	private static void writeLong(long value, DataOutput out) {
+		try {
+			for (int i = 0; i < 8 && (value & ~0x7FL) != 0; i++) {
+				out.writeByte(0x80 | ((int) value & 0x7F));
+				value >>>= 7;
+			}
+			out.writeByte((int) value);
+		} catch (IOException e) {
+			throw new RaftException("Fail to write", e);
+		}
+	}
+
+	private static long readLong(DataInput in) {
+		try {
+			long v = in.readByte(); if (v >= 0) return v;
+			if ((v ^= (long) in.readByte() << 7)  <  0L) return v ^ 0xFFFFFFFFFFFFFF80L;
+			if ((v ^= (long) in.readByte() << 14) >= 0L) return v ^ 0x0000000000003F80L;
+			if ((v ^= (long) in.readByte() << 21) <  0L) return v ^ 0xFFFFFFFFFFE03F80L;
+			if ((v ^= (long) in.readByte() << 28) >= 0L) return v ^ 0x000000000FE03F80L;
+			if ((v ^= (long) in.readByte() << 35) <  0L) return v ^ 0xFFFFFFF80FE03F80L;
+			if ((v ^= (long) in.readByte() << 42) >= 0L) return v ^ 0x000003F80FE03F80L;
+			if ((v ^= (long) in.readByte() << 49) <  0L) return v ^ 0xFFFE03F80FE03F80L;
+			return v ^ (long) in.readByte() << 56 ^ 0x00FE03F80FE03F80L;
+		} catch (IOException e) {
+			throw new RaftException("Fail to read", e);
+		}
+	}
```

Comment:
Yes, I noticed the read/write address methods in org.jgroups.util.Util, but it will serialize/deserialize UUID with its actual type which is unnecessary for the state machine. Regarding the compressed int and long, I found them as well, but I think these are more efficient that it will be only one byte if the number is less than 128 which is the most case for a list size or a lockId I think. UUID's MSB and LSB on the other hand, almost every byte is not 0, so I write them directly. Those methods are inspired by protobuf, I can change them to the compressed version If you insist.

Reply:
Sounds good, no problems.